nav2_graceful_controller/src/graceful_controller.cpp

```diff
@@ -199,24 +153,108 @@ geometry_msgs::msg::TwistStamped GracefulController::computeVelocityCommands(
+      cmd_vel.twist = rotateToTarget(angle_to_goal);
+      return cmd_vel;
+    }
+    // Else, fall through and see if we should follow control law longer
   }
-  local_plan.header = transformed_plan.header;
-  local_plan_pub_->publish(local_plan);
 
-  return cmd_vel;
+  // Precompute distance to candidate poses
+  std::vector<double> target_distances;
+  computeDistanceAlongPath(transformed_plan.poses, target_distances);
+
+  // Work back from the end of plan to find valid target pose
+  for (int i = transformed_plan.poses.size() - 1; i >= 0; --i) {
+    // Underlying control law needs a single target pose, which should:
+    //  * Be as far away as possible from the robot (for smoothness)
+    //  * But no further than the max_lookahed_ distance
+    //  * Be feasible to reach in a collision free manner
+    geometry_msgs::msg::PoseStamped target_pose = transformed_plan.poses[i];
+    double dist_to_target = target_distances[i];
+
+    // Continue if target_pose is too far away from robot
+    if (dist_to_target > params_->max_lookahead) {continue;}
+
+    if (dist_to_goal < params_->max_lookahead) {
+      if (params_->prefer_final_rotation) {
+        // Avoid unstability and big sweeping turns at the end of paths by
+        // ignoring final heading
+        double yaw = std::atan2(target_pose.pose.position.y, target_pose.pose.position.x);
+        target_pose.pose.orientation = nav2_util::geometry_utils::orientationAroundZAxis(yaw);
+      }
+    } else if (dist_to_target < params_->min_lookahead) {
+      // Make sure target is far enough away to avoid instability
+      break;
+    }
+
+    // Flip the orientation of the motion target if the robot is moving backwards
+    bool reversing = false;
+    if (params_->allow_backward && target_pose.pose.position.x < 0.0) {
+      reversing = true;
+      target_pose.pose.orientation = nav2_util::geometry_utils::orientationAroundZAxis(
+        tf2::getYaw(target_pose.pose.orientation) + M_PI);
+    }
+
+    // Actually simulate our path
+    nav_msgs::msg::Path local_plan;
+    if (simulateTrajectory(target_pose, costmap_transform, local_plan, cmd_vel, reversing)) {
+      // Successfully simulated to target_pose - compute velocity at this moment
+      // Publish the selected target_pose
+      auto motion_target_point = nav2_graceful_controller::createMotionTargetMsg(target_pose);
+      motion_target_pub_->publish(motion_target_point);
+      // Publish marker for slowdown radius around motion target for debugging / visualization
+      auto slowdown_marker = nav2_graceful_controller::createSlowdownMarker(
+        target_pose, params_->slowdown_radius);
+      slowdown_pub_->publish(slowdown_marker);
+      // Publish the local plan
+      local_plan.header = transformed_plan.header;
+      local_plan_pub_->publish(local_plan);
+      // Successfully found velocity command
+      return cmd_vel;
+    }
+  }
+
+  throw nav2_core::NoValidControl("Collision detected in trajectory");
```

Comment:
Or also that there is no path further away than `params_->min_lookahead`. I don't see how that is gracefully handled for short paths (or very long lookaheads)

Reply:
Once the goal is < max_lookahead away, we stop enforcing the requirement that target distance > min_lookahead

Reply:
I don't see that in play - maybe I'm missing something. What I think I see is that:
- when `dist_to_target > params_->max_lookahead`, we pass as we're too far
- when `dist_to_goal < params_->max_lookahead` we go through the loop, maybe adjusting the orientation (maybe not)
- When we don't find anything in that regime that works, in `dist_to_target < params_->min_lookahead`, we break the loop and throw an exception

If there's collisions in the full min_lookahead to max_lookahead range and have path options in that range, we throw. This makes sense.

However, if that path itself is shorter than min_lookahead, I don't see any handling. The `goal_dist_tolerance_` and `min_lookahead` are not related or have a requirement on their comparative relationship that would cause a previous rotate-to-heading action to strictly occur. 

Instead, I see on L206 (`double dist_to_target = target_distances[i];`) we find the distance, then its less than the lookahead minimum (`dist_to_target < params_->min_lookahead`) and then `break;`s into the throw.

Reply:
Maybe I don't understand your question here - what are you expecting for "gracefully handling" short paths?

* L209 - as you noted, ignores targets that are too far away
* L211 - IF dist_to_goal < max_lookahead then apply optionally prefer_final_rotation and try to simulate to the target (if this fails, we iterate back through the loop and try the next closest target)
* L218 - ELSE (AKA if dist_to_goal >= max_lookahead) check if (dist_to_target < min_lookahead) - in this case, we have lots of path left, but everything between min and max_lookahead is not simulatable. I have a `break` here rather than a `continue` because every other target should be closer... and **as I write this, I'm realizing that may not be true if we have a path that doubles back onto itself. Is your comment actually suggesting that we just change the `break` on L220 to a `continue`?**

`min_lookahead` should be less than `max_lookahead` (we can add some checking parameter settings to ensure this - but logically, setting min > max is just not going to work).

`dist_to_target` should also always be <= `dist_to_goal` (which we compute on L160) - because at worst case our target is the end of the path.

`goal_dist_tolerance_` isn't involved in this part of the code - so I'm not sure what that comment is about.

Reply:
What I mean is what happens when we have a very short path? If we have a minimum lookahead of 1m (for some large robot) and we've replanned within 0.7m to the goal, but we have a goal tolerance of 0.3m? 

I think right now, we're throw as the distance is < min lookahead and simply fail to control ever. 

> dist_to_target should also always be <= dist_to_goal (which we compute on L160) - because at worst case our target is the end of the path.

OH **dist_to_target** vs **dist_to_goal**, that totally escaped me in reading it.

Reply:
>  I have a break here rather than a continue because every other target should be closer... and as I write this, I'm realizing that may not be true if we have a path that doubles back onto itself. 

I think that should be fine with the looping over itself, since the distances are path integrated distances in both `computeDistanceAlongPath()` and `calculate_path_length()`.

Reply:
> I think that should be fine with the looping over itself, since the distances are path integrated distances in both `computeDistanceAlongPath()` and `calculate_path_length()`, no?

Ah yes - OK. that part works.